EIPS/eip-ondemand-data.md

```diff
@@ -0,0 +1,152 @@
+}
+```
+
+An oracle provider could create a contract (that might also perform some pre-processing) that would automatically trigger a request for off-chain data as follows:
+
+```solidity
+contract OracleContract is IOracleContract {
+
+  address public constant VERIFIER_CONTRACT = 0x0000;
+  uint public constant STALENESS_TOLERANCE = 86400; // One day
+  mapping(bytes32 => bytes) public latestVerifiedData;
+
+  function fulfillOracleData(bytes calldata oracleQuery, bytes calldata signedOffchainData) public {
+    _verify(signedOffchainData);
+    latestVerifiedData[keccak256(oracleQuery)] = signedOffchainData;
+  }
+
+  function retrieveCrossChainData(uint chainId, address contractAddress, bytes payload) internal returns (bytes) {
+    bytes memory oracleQuery = abi.encode(chainId, contractAddress, payload);
+    (uint timestamp, bytes response) = abi.decode(latestVerifiedData[oracleQuery], (uint, bytes));
+
+    if(timestamp < block.timestamp - STALENESS_TOLERANCE){
+      revert OracleDataRequired(ORACLE_ID, oracleQuery, address(oracleContract));
+    }
+
+    return response;
+  }
+
+  function _verify(bytes memory signedOffchainData) {
+    // ... insert verification code here ...
+  }
+
+}
+```
+
+Now a top-level protocol smart contract could implement a cross-chain function like so:
+
+```solidity
+interface ICrosschainContract {
+  function functionA(uint x) external returns (uint y);
+  function functionB(uint x) external returns (uint y);
+}
+
+contract CrosschainAdder {
+  uint public constant ORACLE_ID = 123;
+  IOracleContract oracleContract = 0x0000;
+
+  function add(uint chainIdA, address contractAddressA, uint chainIdB, address contractAddressB) external returns (uint sum){
+    sum = abi.decode(oracleContract.retrieveCrossChainData(chainIdA, contractAddressA, abi.encodeWithSelector(ICrosschainContract.functionA.selector,1)), (uint)) + abi.decode(oracleContract.retrieveCrossChainData(chainIdB, contractAddressB, abi.encodeWithSelector(ICrosschainContract.functionB.selector,2)),(uint));
+  }
+}
+```
+
+Note that the `add` function can simply call the cross-chain function on the Oracle Contract. The developer of the `CrosschainAdder` contract does not need to be concerned with the implementation of this standard.
+
+Cross-chain functions like this could also be leveraged to avoid O(n) (and greater) loops on-chain. For example, `chainIdA` and `chainIdB` could reference the same chain that the `CrosschainAdder` contract is deployed on with `functionA` and `functionB` as view functions with computationally intensive loops.
+
+## Security Considerations
+
+This standard has similar security considerations to [EIP-3668](./eip-3668.md). However, regarding HTTP requests, it is more secure as the client is not expected to query an arbitrary HTTP endpoint specified by the contract. It is instead the responsibility of the client to decide which gateway should be used to connect to the decentralized oracle network specified in the error data.
```

Comment:
I like to see how this is gonna work out practically. i.e: from where the clients get the urls? how clients make sure there is no conflicing id? Is there a registry of some sort? Do clients get the urls upon instantiation?

Reply:
yeah it's probably good to say a word about this. can likely copy the state of affairs for chain ids